### PR TITLE
Remove cast-align disable flags

### PR DESCRIPTION
**Pull request by bugobliterator**
There are some cast-align disable flags being added globally for many builds. Probably not a good idea.

### CODE REVIEW

**Comment by bugobliterator** [comment]
Seems like there are actual errors hiding behind -Wno-cast-align we had enabled for all ChibiOS builds using CAN. @rmackay9 , I wonder if this issue/PR should be part of 4.1

File: libraries/AP_HAL/utility/RingBuffer.h [hunk 1]
```diff
@@ -238,7 +238,10 @@ class ObjectBuffer {
     // !!! Note ObjectBuffer_TS is a duplicate of this, update in both places !!!
     const T *readptr(uint32_t &n) {
         uint32_t avail_bytes = 0;
+        #pragma GCC diagnostic push
+        #pragma GCC diagnostic ignored "-Wcast-align"
```
Comment (khancyr) [inline libraries/AP_HAL/utility/RingBuffer.h:242]: you should do the same L414 as said in the comment

Comment (bugobliterator) [reply libraries/AP_HAL/utility/RingBuffer.h:242]: ha! nice catch

**Review by khancyr — COMMENTED** [review]
Isn't that a place where reinterpret_cast/dynamic_cast would be a good use ?

**Comment by bugobliterator** [comment]
> Aren't those places where reinterpret_cast/dynamic_cast would be a good use ?
you mean instead of using unions? if so using reinterpret_cast will not solve the issue, and most likely compiler will still throw an error. reinterpret_cast doesn't ensure alignment, It just tells compiler to "reinterpret" the pointer as another type.

File: libraries/AP_BLHeli/AP_BLHeli.cpp [hunk 2]
```diff
@@ -743,7 +743,7 @@ bool AP_BLHeli::BL_ConnectEx(void)
 
     blheli.interface_mode[blheli.chan] = 0;
 
-    uint16_t *devword = (uint16_t *)blheli.deviceInfo[blheli.chan];
+    uint16_t *devword = (uint16_t *)blheli.hw_deviceInfo[blheli.chan];
```
Comment (tridge) [inline libraries/AP_BLHeli/AP_BLHeli.cpp:746]: doesn't look right, would need to be chan/2, but better to use functions that pull by byte

File: libraries/AP_CANManager/AP_CANTester_KDECAN.cpp [hunk 3]
```diff
@@ -150,17 +150,18 @@ void AP_CANTester_KDECAN::loop(void)
...
                 }
                 case TELEMETRY_OBJ_ADDR: {
-                    uint8_t data[8] {};
-                    *((be16_t*) &data[0]) = htobe16(get_random16());
-                    *((be16_t*) &data[2]) = htobe16(get_random16());
-                    *((be16_t*) &data[4]) = htobe16(get_random16());
-                    data[6] = uint8_t(float(rand()) / RAND_MAX * 40.0f + 15);
+                    uint16_t data[4] {};
+
+                    data[0] = htobe16(get_random16());
```
Comment (tridge) [inline libraries/AP_CANManager/AP_CANTester_KDECAN.cpp:161]: byte order on random bytes doesn't matter much

File: libraries/AP_HAL/CANIface.h [hunk 4]
```diff
@@ -35,7 +35,13 @@ struct AP_HAL::CANFrame {
     static const uint8_t MaxDataLen = 8;
 
     uint32_t id;                ///< CAN ID with flags (above)
-    uint8_t data[MaxDataLen];
+    union {
+        uint8_t data[MaxDataLen];
+        uint16_t hw_data[MaxDataLen/2];
+        uint32_t w_data[MaxDataLen/4];
+        uint64_t dw_data[MaxDataLen/8];
```
Comment (tridge) [inline libraries/AP_HAL/CANIface.h:42]: this makes CANFrame 4 bytes larger, maybe avoid dw_data?

File: libraries/AP_EFI/AP_EFI_NWPMU.h [hunk 5]
```diff
@@ -79,6 +79,7 @@ class AP_EFI_NWPMU : public CANSensor, public AP_EFI_Backend {
         uint16_t map;    // 0.01 per bit
         uint16_t lambda; // 0.01 per bit
         uint8_t pressure_type; // bit 1 is set if kPa, otherwise is PSI
+        uint8_t unused;   // unused
```
Comment (tridge) [inline libraries/AP_EFI/AP_EFI_NWPMU.h:82]: i don't think these are needed

Comment (tridge) [reply libraries/AP_EFI/AP_EFI_NWPMU.h:82]: probably better to just memcpy and declare the structure on the stack

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_BLHeli/AP_BLHeli.cpp (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_BLHeli/AP_BLHeli.cpp]: done

**Review by tridge — APPROVED** [review]

**Pull request merged**